storage/cache/subtree_cache.go

```diff
@@ -259,55 +214,49 @@ func (s *SubtreeCache) getNodeHash(id compact.NodeID, getSubtree GetSubtreeFunc)
 	// have 8 significant suffix bits.
 	sfxKey := sx.String()
 	if int32(sx.Bits()) == c.Depth {
-		nh = c.Leaves[sfxKey]
-	} else {
-		nh = c.InternalNodes[sfxKey]
+		return c.Leaves[sfxKey], nil
 	}
-	return nh, nil
+	return c.InternalNodes[sfxKey], nil
 }
 
-// SetNodeHash sets a node hash in the cache.
-func (s *SubtreeCache) SetNodeHash(id compact.NodeID, h []byte, getSubtree GetSubtreeFunc) error {
-	subID, sx := splitID(id)
-	c := s.getCachedSubtree(subID)
-	if c == nil {
-		// TODO(al): This is ok, IFF *all* leaves in the subtree are being set,
-		// verify that this is the case when it happens.
-		// For now, just read from storage if we don't already have it.
-		glog.V(1).Infof("attempting to write to unread subtree for %+v, reading now", id)
-		if _, err := s.getNodeHash(id, getSubtree); err != nil {
-			return err
-		}
-		// There must be a subtree present in the cache now, even if storage didn't have anything for us.
-		c = s.getCachedSubtree(subID)
-		if c == nil {
-			return fmt.Errorf("internal error, subtree cache for %+v is nil after a read attempt", id)
-		}
+// SetNodes sets hashes for the given nodes in the cache.
+func (s *SubtreeCache) SetNodes(nodes []tree.Node, getSubtrees GetSubtreesFunc) error {
+	ids := make([]compact.NodeID, len(nodes))
+	for i, n := range nodes {
+		ids[i] = n.ID
 	}
-	if c.Prefix == nil {
-		return fmt.Errorf("nil prefix for %+v (key %v)", id, subID)
+	notFound, err := s.preload(ids, getSubtrees)
+	if err != nil {
+		return err
 	}
-	// Determine whether we're being asked to store a leaf node, or an internal
-	// node, and store it accordingly.
-	sfxKey := sx.String()
-	if int32(sx.Bits()) == c.Depth {
-		// If the value being set is identical to the one we read from storage, then
-		// leave the cache state alone, and return.  This will prevent a write (and
-		// subtree revision bump) for identical data.
-		if bytes.Equal(c.Leaves[sfxKey], h) {
-			return nil
+	for _, id := range notFound {
+		c := newEmptyTile([]byte(id))
+		s.subtrees.Store(string(id), c)
+	}
+
+	for _, n := range nodes {
+		subID, sx := splitID(n.ID)
+		c := s.getCachedSubtree(subID)
+		if c == nil {
+			return fmt.Errorf("tile %x not found", subID)
 		}
-		c.Leaves[sfxKey] = h
-	} else {
-		// If the value being set is identical to the one we read from storage, then
-		// leave the cache state alone, and return.  This will prevent a write (and
-		// subtree revision bump) for identical data.
-		if bytes.Equal(c.InternalNodes[sfxKey], h) {
-			return nil
+
+		// Determine whether we're being asked to store a leaf node, or an internal
+		// node, and store it accordingly. If the value being set is identical to
+		// the one we read from storage, then leave the cache state alone. This
+		// will prevent a write (and subtree revision bump) for identical data.
+		sfxKey := sx.String()
```

Comment:
This section below is hard to evaluate for correctness.

Reply:
Refactored a bit. Does it look better now?

Reply:
The comments help.